File: src/json_prob_parser/heuristic.py
from __future__ import annotations

from typing import List, Tuple

from .types import RepairAction, RepairOptions


def _fix_smart_quotes(text: str) -> Tuple[str, List[RepairAction]]:
    # Common in copy/paste from docs or rich text.
    replacements = {
        "\u201c": '"',  # “
        "\u201d": '"',  # ”
        "\u2018": "'",  # ‘
        "\u2019": "'",  # ’
    }
    if not any(ch in text for ch in replacements):
        return text, []
    out = text.translate(str.maketrans(replacements))
    # Base cost only (we avoid per-char cost to keep v1 simple and deterministic).
    return out, [RepairAction(op="fix_smart_quotes", cost_delta=0.7)]


def _strip_comments(text: str) -> Tuple[str, List[RepairAction]]:
    out: list[str] = []
    repairs: list[RepairAction] = []
    i = 0
    in_string = False
    escape = False
    while i < len(text):
        ch = text[i]
        if in_string:
            out.append(ch)
            if escape:
                escape = False
            elif ch == "\\":
                escape = True
            elif ch == '"':
                in_string = False
            i += 1
            continue

        if ch == '"':
            in_string = True
            out.append(ch)
            i += 1
            continue

        if ch == "/" and i + 1 < len(text) and text[i + 1] == "/":
            start = i
            i += 2
            while i < len(text) and text[i] not in "\r\n":
                i += 1
            repairs.append(RepairAction(op="strip_line_comment", span=(start, i), cost_delta=0.4))
            continue

        if ch == "/" and i + 1 < len(text) and text[i + 1] == "*":
            start = i
            i += 2
            while i + 1 < len(text) and not (text[i] == "*" and text[i + 1] == "/"):
                i += 1
            i = min(len(text), i + 2)
            repairs.append(RepairAction(op="strip_block_comment", span=(start, i), cost_delta=0.6))
            continue

        out.append(ch)
        i += 1

    return "".join(out), repairs


def _normalize_python_literals(text: str) -> Tuple[str, List[RepairAction]]:
    out: list[str] = []
    repairs: list[RepairAction] = []
    i = 0
    in_string = False
    escape = False
    mapping = {"True": "true", "False": "false", "None": "null", "undefined": "null"}
    while i < len(text):
        ch = text[i]
        if in_string:
            out.append(ch)
            if escape:
                escape = False
            elif ch == "\\":
                escape = True
            elif ch == '"':
                in_string = False
            i += 1
            continue

        if ch == '"':
            in_string = True
            out.append(ch)
            i += 1
            continue

        if ch.isalpha() or ch == "_":
            start = i
            i += 1
            while i < len(text) and (text[i].isalnum() or text[i] == "_"):
                i += 1
            word = text[start:i]
            repl = mapping.get(word)
            if repl is not None:
                out.append(repl)
                repairs.append(
                    RepairAction(op="map_python_literal", span=(start, i), cost_delta=0.4, note=f"{word}->{repl}")
                )
            else:
                out.append(word)
            continue

        out.append(ch)
        i += 1

    return "".join(out), repairs


def _remove_trailing_commas(text: str) -> Tuple[str, List[RepairAction]]:
    out: list[str] = []
    repairs: list[RepairAction] = []
    i = 0
    in_string = False
    escape = False
    while i < len(text):
        ch = text[i]
        if in_string:
            out.append(ch)
            if escape:
                escape = False
            elif ch == "\\":
                escape = True
            elif ch == '"':
                in_string = False
            i += 1
            continue

        if ch == '"':
            in_string = True
            out.append(ch)
            i += 1
            continue

        if ch == ",":
            j = i + 1
            while j < len(text) and text[j].isspace():
                j += 1
            if j >= len(text) or text[j] in ("}", "]"):
                repairs.append(RepairAction(op="remove_trailing_comma", at=i, cost_delta=0.2))
                i += 1
                continue

        out.append(ch)
        i += 1

    return "".join(out), repairs


def _append_missing_closers(text: str) -> Tuple[str, List[RepairAction]]:
    repairs: list[RepairAction] = []
    in_string = False
    escape = False
    depth_brace = 0
    depth_bracket = 0
    for ch in text:
        if in_string:
            if escape:
                escape = False
            elif ch == "\\":
                escape = True
            elif ch == '"':
                in_string = False
            continue
        if ch == '"':
            in_string = True
            continue
        if ch == "{":
            depth_brace += 1
        elif ch == "}":
            depth_brace -= 1
        elif ch == "[":
            depth_bracket += 1
        elif ch == "]":
            depth_bracket -= 1

    out = text
    if in_string:
        out += '"'
        repairs.append(RepairAction(op="close_open_string", at=len(text), cost_delta=3.0))

    if depth_brace > 0 or depth_bracket > 0:
        out += ("]" * max(0, depth_bracket)) + ("}" * max(0, depth_brace))
        repairs.append(
            RepairAction(
                op="close_containers",
                at=len(text),
                cost_delta=0.5 * (max(0, depth_brace) + max(0, depth_bracket)),
                note=f"brace={depth_brace}, bracket={depth_bracket}",
            )
        )

    return out, repairs


def heuristic_repair(extracted_text: str, opt: RepairOptions) -> Tuple[str, List[RepairAction]]:
    text = extracted_text
    repairs: list[RepairAction] = []

    text2, acts = _fix_smart_quotes(text)
    if text2 != text:
        text = text2
        repairs.extend(acts)

    if opt.allow_comments:
        text2, acts = _strip_comments(text)
        if text2 != text:
            text = text2
            repairs.extend(acts)

    if opt.allow_python_literals:
        text2, acts = _normalize_python_literals(text)
        if text2 != text:
            text = text2
            repairs.extend(acts)

    text2, acts = _remove_trailing_commas(text)
    if text2 != text:
        text = text2
        repairs.extend(acts)

    text2, acts = _append_missing_closers(text)
    if text2 != text:
        text = text2
        repairs.extend(acts)

    return text, repairs

File: rust/src/heuristic.rs
use crate::types::{RepairAction, RepairOptions};

fn fix_smart_quotes(text: &str) -> (String, Vec<RepairAction>) {
    let mut out = String::with_capacity(text.len());
    let mut changed = false;
    for ch in text.chars() {
        match ch {
            '\u{201C}' | '\u{201D}' => {
                out.push('"');
                changed = true;
            }
            '\u{2018}' | '\u{2019}' => {
                out.push('\'');
                changed = true;
            }
            _ => out.push(ch),
        }
    }
    if changed {
        (out, vec![RepairAction::new("fix_smart_quotes", 0.7)])
    } else {
        (text.to_string(), vec![])
    }
}

fn strip_comments(text: &str) -> (String, Vec<RepairAction>) {
    let bytes = text.as_bytes();
    let mut out: Vec<u8> = Vec::with_capacity(bytes.len());
    let mut repairs = Vec::new();
    let mut i: usize = 0;
    let mut in_string = false;
    let mut escape = false;
    while i < bytes.len() {
        let ch = bytes[i];
        if in_string {
            out.push(ch);
            if escape {
                escape = false;
            } else if ch == b'\\' {
                escape = true;
            } else if ch == b'"' {
                in_string = false;
            }
            i += 1;
            continue;
        }

        if ch == b'"' {
            in_string = true;
            out.push(ch);
            i += 1;
            continue;
        }

        if ch == b'/' && i + 1 < bytes.len() && bytes[i + 1] == b'/' {
            let start = i;
            i += 2;
            while i < bytes.len() && bytes[i] != b'\n' && bytes[i] != b'\r' {
                i += 1;
            }
            let mut a = RepairAction::new("strip_line_comment", 0.4);
            a.span = Some((start, i));
            repairs.push(a);
            continue;
        }

        if ch == b'/' && i + 1 < bytes.len() && bytes[i + 1] == b'*' {
            let start = i;
            i += 2;
            while i + 1 < bytes.len() && !(bytes[i] == b'*' && bytes[i + 1] == b'/') {
                i += 1;
            }
            i = (i + 2).min(bytes.len());
            let mut a = RepairAction::new("strip_block_comment", 0.6);
            a.span = Some((start, i));
            repairs.push(a);
            continue;
        }

        out.push(ch);
        i += 1;
    }
    (String::from_utf8_lossy(&out).to_string(), repairs)
}

fn normalize_python_literals(text: &str) -> (String, Vec<RepairAction>) {
    let bytes = text.as_bytes();
    let mut out: Vec<u8> = Vec::with_capacity(bytes.len());
    let mut repairs = Vec::new();
    let mut i: usize = 0;
    let mut in_string = false;
    let mut escape = false;
    while i < bytes.len() {
        let ch = bytes[i];
        if in_string {
            out.push(ch);
            if escape {
                escape = false;
            } else if ch == b'\\' {
                escape = true;
            } else if ch == b'"' {
                in_string = false;
            }
            i += 1;
            continue;
        }

        if ch == b'"' {
            in_string = true;
            out.push(ch);
            i += 1;
            continue;
        }

        if (ch as char).is_ascii_alphabetic() || ch == b'_' {
            let start = i;
            i += 1;
            while i < bytes.len() && (((bytes[i] as char).is_ascii_alphanumeric()) || bytes[i] == b'_') {
                i += 1;
            }
            let word = &text[start..i];
            let mapped = match word {
                "True" => Some("true"),
                "False" => Some("false"),
                "None" => Some("null"),
                "undefined" => Some("null"),
                _ => None,
            };
            if let Some(m) = mapped {
                out.extend_from_slice(m.as_bytes());
                let mut a = RepairAction::new("map_python_literal", 0.4);
                a.span = Some((start, i));
                a.note = Some(format!("{word}->{m}"));
                repairs.push(a);
            } else {
                out.extend_from_slice(word.as_bytes());
            }
            continue;
        }

        out.push(ch);
        i += 1;
    }
    (String::from_utf8_lossy(&out).to_string(), repairs)
}

fn is_ws(b: u8) -> bool {
    matches!(b, b' ' | b'\n' | b'\r' | b'\t')
}

fn remove_trailing_commas(text: &str) -> (String, Vec<RepairAction>) {
    let bytes = text.as_bytes();
    let mut out: Vec<u8> = Vec::with_capacity(bytes.len());
    let mut repairs = Vec::new();
    let mut i: usize = 0;
    let mut in_string = false;
    let mut escape = false;
    while i < bytes.len() {
        let ch = bytes[i];
        if in_string {
            out.push(ch);
            if escape {
                escape = false;
            } else if ch == b'\\' {
                escape = true;
            } else if ch == b'"' {
                in_string = false;
            }
            i += 1;
            continue;
        }

        if ch == b'"' {
            in_string = true;
            out.push(ch);
            i += 1;
            continue;
        }

        if ch == b',' {
            let mut j = i + 1;
            while j < bytes.len() && is_ws(bytes[j]) {
                j += 1;
            }
            if j >= bytes.len() || bytes[j] == b'}' || bytes[j] == b']' {
                let mut a = RepairAction::new("remove_trailing_comma", 0.2);
                a.at = Some(i);
                repairs.push(a);
                i += 1;
                continue;
            }
        }

        out.push(ch);
        i += 1;
    }
    (String::from_utf8_lossy(&out).to_string(), repairs)
}

fn append_missing_closers(text: &str) -> (String, Vec<RepairAction>) {
    let bytes = text.as_bytes();
    let mut in_string = false;
    let mut escape = false;
    let mut depth_brace: i64 = 0;
    let mut depth_bracket: i64 = 0;
    let mut i: usize = 0;
    while i < bytes.len() {
        let ch = bytes[i];
        if in_string {
            if escape {
                escape = false;
            } else if ch == b'\\' {
                escape = true;
            } else if ch == b'"' {
                in_string = false;
            }
            i += 1;
            continue;
        }
        if ch == b'"' {
            in_string = true;
            i += 1;
            continue;
        }
        match ch {
            b'{' => depth_brace += 1,
            b'}' => depth_brace -= 1,
            b'[' => depth_bracket += 1,
            b']' => depth_bracket -= 1,
            _ => {}
        }
        i += 1;
    }

    let mut out = text.to_string();
    let mut repairs = Vec::new();
    if in_string {
        out.push('"');
        let mut a = RepairAction::new("close_open_string", 3.0);
        a.at = Some(text.len());
        repairs.push(a);
    }

    let b = depth_brace.max(0) as usize;
    let a = depth_bracket.max(0) as usize;
    if b > 0 || a > 0 {
        out.push_str(&"]".repeat(a));
        out.push_str(&"}".repeat(b));
        let mut act = RepairAction::new("close_containers", 0.5 * ((a + b) as f64));
        act.at = Some(text.len());
        act.note = Some(format!("brace={b}, bracket={a}"));
        repairs.push(act);
    }
    (out, repairs)
}

pub fn heuristic_repair(extracted_text: &str, opt: &RepairOptions) -> (String, Vec<RepairAction>) {
    let mut text = extracted_text.to_string();
    let mut repairs: Vec<RepairAction> = Vec::new();

    let (t2, r2) = fix_smart_quotes(&text);
    if t2 != text {
        text = t2;
        repairs.extend(r2);
    }

    if opt.allow_comments {
        let (t2, r2) = strip_comments(&text);
        if t2 != text {
            text = t2;
            repairs.extend(r2);
        }
    }

    if opt.allow_python_literals {
        let (t2, r2) = normalize_python_literals(&text);
        if t2 != text {
            text = t2;
            repairs.extend(r2);
        }
    }

    let (t2, r2) = remove_trailing_commas(&text);
    if t2 != text {
        text = t2;
        repairs.extend(r2);
    }

    let (t2, r2) = append_missing_closers(&text);
    if t2 != text {
        text = t2;
        repairs.extend(r2);
    }

    (text, repairs)
}

